Create orka-ipsw-setup.sh

celanthe opened the pull request:
adds a shell script for setting up a MacOS VM sourced from IPSW that can be deployed using Orka, and adds supporting documentation.

celanthe committed:
Create orka-ipsw-setup-tahoe.sh

adds a shell script for creating an image via IPSW that can be deployed using Orka

celanthe committed:
Create vm-setup-script-usage.md

adds supporting documentation for using Orka VM setup script

celanthe committed:
update VM tools pkg version

bump orka vm tools version to 3.5.0(latest)

celanthe committed:
Update vm-setup-script-usage.md

bump version of orka VM tools mentioned in docs to 3.5.0(latest)

celanthe committed:
add logging functions

add missing log functions

GerryMandell commented on vm-setup-script-usage.md:
````diff
@@ -0,0 +1,87 @@
+
+From your local machine:
+
+- ```bashscp orka-ipsw-setup-tahoe.sh user@vm-ip:/Users/user/```
````
small nit, spacing on this example command between `bash` and `scp`

celanthe replied:
fixed this! :)

celanthe committed:
add color settings

adds error color settings to script

celanthe committed:
fix typo

add a space between bash and scpt

celanthe committed:
add VM tools pkg_name

add vm tools pkg name to output

celanthe committed:
Update vm-setup-script-usage.md

adds install instructions on MacOS Sequoia

celanthe committed:
Update vm-setup-script-usage.md

add tahoe VM delay note to README

celanthe committed:
Update vm-setup-script-usage.md

add Sonoma setup instructions to README

celanthe committed:
Rename orka-ipsw-setup-tahoe.sh to orka-ipsw-setup.sh

rename script

celanthe committed:
Update vm-setup-script-usage.md

update name of script in docs

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,269 @@
+        -clientopts -setvncpw -vncpw admin -restart -agent -privs -all
+    
+    # Alternative method using launchctl
+    sudo launchctl load -w /System/Library/LaunchDaemons/com.apple.screensharing.plist 2>/dev/null || true
```
Do we need to run both this and the above kickstart command?

I believe the first command (kickstart) will enable Remote Management, which also enables screen sharing. This second command only enables screen sharing

Usually, I believe we only enable Screen Sharing, so the second command should suffice

Also - this is somewhat `nit` - but `launchctl load` is deprecated. I would favor using `launchctl bootstrap` if possible to modernize this. However - even the [AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/connect-to-mac-instance.html#mac-instance-vnc) instruct doing it this way, so it is just a suggestion

celanthe replied:
Thanks, fixed this up in a new commit.

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,269 @@
+    log "Enabling Remote Login (SSH)..."
+    
+    # Create admin user if one doesn't already exist
+    if ! dscl . -read /Users/admin &>/dev/null; then
```
Don't we just care that there is an existing administrator user? Not that there is a user explicitly named `admin`? (when you go through the initial macOS setup wizard - you create the first administrator user and can name it whatever you like)

This logic seems unnecessary. And we should not "force" the creation of a new administrator user account - this is pretty undesirable

jasoncdavis0 replied:
Today, we provide admin/admin as the un/pw to log in to the instance. Are you suggesting we change that and give the user some more control over admin user creation?

celanthe replied:
Fair point, removed this.

spikeburton replied:
@jasoncdavis0 this is the case for the images that we provide, yes.

But isn't the idea of this script for someone to create a new VM from an IPSW themselves, which includes going through the initial macOS setup wizard, and then running this script to configure the VM to be able to run with Orka seamlessly?

Or do I misunderstand the intent?

We have no hard requirement for a user called "admin". The administrator account can be called whatever you like, and this is created as part of the initial macOS setup wizard. Chances are, most users will not want to create an administrator account with the credentials of admin/admin

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,269 @@
+        log "Admin user created with password 'admin'"
+    else
+        log "Admin user already exists, updating password..."
+        sudo dscl . -passwd /Users/admin admin
```
Why would we want to update the password here to "admin"? Or - why do we need to update the password at all here?

celanthe replied:
When we build images for orka-images the username/password we set is admin/admin and that is what we reference in [our documentation](https://support.macstadium.com/hc/en-us/articles/28347375974555-Orka3-CLI-Quick-Start), but we don't have to update it if it is already set.

spikeburton replied:
Yes, but assuming that the user is running this after having gone through the initial macOS setup wizard, the administrator user and password will be created already. Right? Or do I misunderstand the intent

celanthe replied:
Yes that's true, I removed this line in a new commit.

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,269 @@
+    local filevault_status=$(sudo fdesetup status | head -1)
+    
+    if [[ "$filevault_status" == "FileVault is On." ]]; then
+        warn "FileVault is currently enabled. For Tahoe, it should be disabled."
```
Is there a specific reason that FileVault should be disabled? It will impact the disk I/O performance for sure (and users should be aware of this). But is it a hard requirement? It is not uncommon to want it enabled

jasoncdavis0 replied:
This one was me - It was not previously listed in manual steps, but I thought it needed to be disabled. It is not a hard product requirement - possibly instead a misunderstanding.

celanthe replied:
We recommend to disable Filevault for [on-prem deployments](https://support.macstadium.com/hc/en-us/articles/40640583440539-Customer-Environment-Tips).

spikeburton replied:
Got it. Yeah, when you go through the initial macOS setup wizard on Tahoe, you're prompted whether to enable FileVault or not. For the stock images that we provide we do not enable it (mainly for performance reasons) - if folks want to enable it though, I do not think we should discourage them from doing so

spikeburton replied:
About bare metal deployments - yes, but this is different. The reason is that it is difficult for a DCT to troubleshoot the machine if they cannot access it due to FileVault. A VM does not present the same concerns

celanthe replied:
I agree, as of now, the script doesn't disable filevault, it checks if it is disabled, and prompts to disable it. I can re-word this or remove it entirely.

celanthe replied:
removed this in the most recent commit

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,269 @@
+install_orka_vm_tools() {
+    log "Installing Orka VM Tools..."
+    
+    local pkg_url="https://orka-tools.s3.amazonaws.com/orka-vm-tools/official/3.5.0/orka-vm-tools.pkg"
```
I would parameterize the VM tools version. Otherwise, we need to update the script every time we update the VM tools. If we want to set the default to `3.5.0` this should be OK, but users should be able to override this when running the script

celanthe replied:
Done!

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,269 @@
+setup_sys_daemon() {
+    log "Setting up system daemon..."
+    
+    local script_url="https://raw.githubusercontent.com/macstadium/packer-plugin-macstadium-orka/main/guest-scripts/setup-sys-daemon.sh"
```
One idea - do we need to download the script? We could just inline it here.

Note - it is just an idea. An advantage of keeping it as a separate script is it is a single source of truth to maintain if it is used elsewhere

celanthe replied:
I am going to keep it a separate script for now. :)

spikeburton commented on setup/vm-setup-script-usage.md:
```diff
@@ -0,0 +1,121 @@
+- Close any important applications (they will be terminated during cleanup)
+- Save any work in progress (the VM will reboot automatically)
+
+### Installation:
```
I would only have a single method of installation. It is much simpler. If you want to download the script to inspect it first, you can.

Homebrew already has a good example of this that is a one-liner right on the homepage: https://brew.sh/

celanthe replied:
Sure thing, updated this.

spikeburton commented on vm-setup-script-usage.md:
````diff
@@ -0,0 +1,121 @@
+
+#### Method 2: Download via GitHub
+
+- ```bashcurl -fsSL https://github.com/macstadium/orka-images/orka-ipsw-setup-tahoe.sh -o orka-ipsw-setup.sh```
````
Note - you can use ```` ```bash ```` for a multiline code block. You do not need the backticks on each line.

I do not think we need several commands here however. I believe this can just be a one-liner:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/orka-ipsw-setup-tahoe.sh)"
```

celanthe replied:
Thanks, updated

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,269 @@
+trap 'error "Script interrupted by user"' INT TERM
+
+# Run main function
+main "$@"
```
Why `$@`? It does not look like we expect any arguments to be passed to `main`. This is used to pass all arguments passed to the script to another script or function

celanthe replied:
Removed this in a new commit

celanthe committed:
Update orka-ipsw-setup.sh

update script to parameterize orka VM tools version, remove unnecessary admin configurations, and duplicate screen share mention.

celanthe committed:
Update vm-setup-script-usage.md

update installation instructions and fix formatting

celanthe committed:
Update orka-ipsw-setup.sh

remove unnecessary step from script

celanthe committed:
Update vm-setup-script-usage.md

Update README to reflect updated install process and filevault permissions check.

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,245 @@
+    log "Enabling Screen Sharing..."
+    
+    # Enable Screen Sharing via System Preferences
+    sudo /System/Library/CoreServices/RemoteManagement/ARDAgent.app/Contents/Resources/kickstart \
```
Now we are enabling Remote Management (instead of just screen sharing). Is this what we want?

celanthe replied:
Will update, thanks.

celanthe replied:
updated this in the most recent commit

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,245 @@
+    log "Enabling Remote Login (SSH)..."
+        
+        # Add to admin group
+        sudo dscl . -append /Groups/admin GroupMembership admin
+        sudo dscl . -append /Groups/wheel GroupMembership admin
+        
+        # Create home directory
+        sudo createhomedir -c -u admin
```
This still assumes there is a user called `admin`.

What was the purpose of this logic originally? Is there something specific required in order to enable SSH?

celanthe replied:
updated this in the most recent commit

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,245 @@
+    sudo dseditgroup -o edit -a everyone -t group com.apple.access_ssh
+    
+    # Ensure admin user can SSH
+    sudo dseditgroup -o edit -a admin -t user com.apple.access_ssh
```
Same here - we are still assuming there is a user called `admin`

celanthe replied:
fixed in most recent commit

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,245 @@
+}
+
+# Disable FileVault (When running on Tahoe)
+disable_filevault() {
```
Did we decide to keep the warning for FileVault?

Btw - there is no logic here specific to Tahoe - this warning would show no matter what the guest OS version is. 

But again - do we need this?

celanthe replied:
will delete

spikeburton commented on orka-ipsw-setup.sh:
```diff
@@ -0,0 +1,245 @@
+    # Clean up installer
+    rm -f "$pkg_path"
+    
+    log "Orka VM Tools v3.5.0 installed successfully"
```
Shouldn't we log the correct version? As it is now a variable. Not hardcode v3.5.0 here

celanthe replied:
will fix thanks

spikeburton commented on vm-setup-script-usage.md:
````diff
@@ -0,0 +1,92 @@
+
+#### Download via GitHub
+
+- ```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/orka-ipsw-setup.sh)"```
````
You do not need all 3 lines here. Again - see the Homebrew example I've pointed out several times now :)) 

This single line will download the script with curl and execute it directly with bash.

celanthe replied:
updated this, thanks

spikeburton replied:
Also - this does not render correctly.

When we use triple backticks, the correct way is to add the language hint directly after the opening backticks, and then separate on newlines. Like this:

````
```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/orka-ipsw-setup.sh)"
```
````

Which renders as this:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/orka-ipsw-setup.sh)"
```

<img width="991" height="185" alt="Screenshot 2025-10-02 at 11 28 32 AM" src="https://github.com/user-attachments/assets/2a3fdf60-9c63-4c76-853e-31888b04afe5" />

celanthe committed:
remove unnecessary commands

remove unnecessary commands, streamline others for consistency

celanthe committed:
Update vm-setup-script-usage.md

update bash script to be one line

spikeburton commented on vm-setup-script-usage.md:
````diff
@@ -0,0 +1,92 @@
+- When prompted for Terminal permissions access, click 'Accept'
+- Terminal will need full disk access to run the script. Configure this by going to your System Settings -> Privacy and Security -> Full Disk Access and adding 'Terminal' by clicking the + button and searching for the app in the search bar.
+- Enter your password when prompted during the script installation process
+- When the script shell displays ```sh-3.2#``` type ```exit``` and press ```Enter``` to continue script installation
````
Also - If we do not need multiline formatted text, or do not care about rendering highlighted syntax, then it is common to just use single backticks. For example this should be

````
When the script shell displays `sh-3.2#` type `exit` and press `Enter` to continue script installation
````

Which renders as:

When the script shell displays `sh-3.2#` type `exit` and press `Enter` to continue script installation

celanthe replied:
fixed

spikeburton commented on vm-setup-script-usage.md:
`````diff
@@ -0,0 +1,89 @@
+- When the script shell displays ```sh-3.2#``` type ```exit``` and press ```Enter``` to continue script installation
+- Upon system reboot, re-open the Terminal application and enter ```sudo launchctl list sysctl``` to confirm the Orka sys-daemon script has installed. You should see an output similar to:
+
+````{
`````
This also does not quite render correctly:

<img width="1096" height="570" alt="Screenshot 2025-10-02 at 11 35 44 AM" src="https://github.com/user-attachments/assets/fd25b5bb-b997-433c-bed7-987931b4c31b" />

celanthe replied:
fixed in a new commit

spikeburton commented:
Something else - it is a bit odd to have the script and the additional README at the root level of the repository. Can we place them inside a directory?

And the main README can link to the additional docs for the setup script. This should be mentioned so folks do not have to dig for it

Maybe also we can simplify the names of the files. For example - maybe just `setup.sh` and `README.md`. What do you think?

celanthe committed:
Update vm-setup-script-usage.md

fix formatting

celanthe committed:
Create setup directory

add a directory called 'setup

celanthe committed:
add setup dir, update main README

celanthe committed:
fix: rename files for clarity

celanthe commented:
> Something else - it is a bit odd to have the script and the additional README at the root level of the repository. Can we place them inside a directory?
> 
> And the main README can link to the additional docs for the setup script. This should be mentioned so folks do not have to dig for it
> 
> Maybe also we can simplify the names of the files. For example - maybe just `setup.sh` and `README.md`. What do you think?

sure thing, done in a new commit

celanthe committed:
Update setup.sh

fix variable formatting and add missing step

celanthe committed:
Update setup.sh

update case formatting

spikeburton commented on README.md:
```diff
@@ -34,6 +34,10 @@ spec:
 
+### VM setup from IPSW script
+
+To get started with Orka using an IPSW and Orka Desktop, use the `orka-ipsw-setup.sh` script located in the `setup` directory.
```
The script is just called `setup.sh` now, right?

Also - why not just link here directly to the script readme? (see [here](https://github.com/macstadium/orka-github-actions-integration/blob/41d3b6e69c75921c435ce3991f1a9abc7cff5c6b/README.md?plain=1#L47) for an example of what I mean by this)

celanthe replied:
good catch, done, thanks!

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,87 @@
+
+#### Download via GitHub
+
+- `bash /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/setup/setup.sh)"`
```
This still does not render correctly. I would suggest previewing the render before pushing the changes for the readme.

<img width="1041" height="137" alt="Screenshot 2025-10-02 at 2 30 10 PM" src="https://github.com/user-attachments/assets/b4e4f235-95e0-4d5c-b753-0f5a7f291b82" />

The correct way to write this (if you want syntax highlighting for `bash`) is to use triple backticks.

Like this:

````
```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/setup/setup.sh)"
```
````

This will render as:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/setup/setup.sh)"
```

celanthe replied:
thanks, pushed a new commit

spikeburton commented on setup/README.md:
````diff
@@ -0,0 +1,87 @@
+- When the script shell displays `sh-3.2#` type `exit` and press `Enter` to continue script installation
+- Upon system reboot, re-open the Terminal application and enter `sudo launchctl list sysctl` to confirm the Orka sys-daemon script has installed. You should see an output similar to:
+
+```markdown
````
Somewhat `nit` - but this still renders a bit awkwardly with the trailing `);`

<img width="545" height="290" alt="Screenshot 2025-10-02 at 2 33 12 PM" src="https://github.com/user-attachments/assets/6532ce0d-271a-4fbd-946b-83632dfc4586" />

celanthe replied:
no worries, pushed a new commit

spikeburton reviewed (commented):
@celanthe can we resolve the merge conflict with the README?

celanthe committed:
Merge branch 'main' into feat-add-ipsw-script-1

celanthe committed:
fix: update README formatting

spikeburton commented on setup/README.md:
````diff
@@ -0,0 +1,88 @@
+#### Download via GitHub
+
+```bash
+/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/setup/setup.sh)"
````
```suggestion
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/macstadium/orka-images/HEAD/setup/setup.sh)"
```

I did not catch this before - but after we merge the PR, this should be the correct path

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,217 @@
+enable_remote_login() {
+    log "Enabling Remote Login (SSH)..."
+    
+    local CURRENT_USER="$USER"
+    
+    # Add current user to admin group if not already a member
+    sudo dscl . -append /Groups/admin GroupMembership "$CURRENT_USER" 2>/dev/null || true
+    sudo dscl . -append /Groups/wheel GroupMembership "$CURRENT_USER" 2>/dev/null || true
+    
+    # Ensure home directory exists for current user
+    sudo createhomedir -c -u "$CURRENT_USER" 2>/dev/null || true
```
Do we need all of this logic? I do not think that we do. An administrator user will have all of this already

celanthe replied:
Removed w/most recent commit

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,217 @@
+    sudo systemsetup -setremotelogin on
+    
+    # Ensure current user can SSH
+    sudo dseditgroup -o edit -a "$CURRENT_USER" -t user com.apple.access_ssh 2>/dev/null || true
```
Same with this. Is it required?

celanthe replied:
Removed this per below update

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,217 @@
+    sudo dseditgroup -o edit -a "$CURRENT_USER" -t user com.apple.access_ssh 2>/dev/null || true
+    
+    # Start SSH service
+    sudo launchctl load -w /System/Library/LaunchDaemons/ssh.plist 2>/dev/null || true
```
I would suggest using either `sudo systemsetup -setremotelogin on` or loading the service manually w/ `launchctl` - I believe doing both is redundant.

Also - I am testing this currently.  It is a pain to have to manually grant Full Disk Access to Terminal.app in order to run the `systemsetup` command, so I wonder if just loading the service w/ `launchctl` is sufficient

spikeburton replied:
UPDATE: I can confirm that just running this line is adequate:

```
sudo launchctl load -w /System/Library/LaunchDaemons/ssh.plist 2>/dev/null || true
```

You are able to SSH to the VM just fine. None of the steps for creating a user home directory or adding the user to a group are needed.

The catch - if you want to be able to access certain directories that are protected by Full Disk Access, you do need to enable this through System Preferences. This we can document by going to System Preferences -> General -> Sharing -> Remote Login and toggling "Allow full disk access for remote users" to On

<img width="1107" height="794" alt="Screenshot 2025-10-03 at 1 11 15 PM" src="https://github.com/user-attachments/assets/c71bca6f-79d2-4236-8f27-7a3fd9cfaace" />

It should not be required to grant Full Disk Access to Terminal.app (and some users may not want to do this, as they will want to grant least privilege)

celanthe replied:
TY, addressed this in the docs and the script in the most recent commit

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,217 @@
+    log "Configuring macOS updates for Download Only..."
+    
+    # Set automatic update check to true, but disable automatic installation
+    sudo defaults write /Library/Preferences/com.apple.SoftwareUpdate.plist AutomaticCheckEnabled -bool true
+    sudo defaults write /Library/Preferences/com.apple.SoftwareUpdate.plist AutomaticDownload -bool true
+    sudo defaults write /Library/Preferences/com.apple.SoftwareUpdate.plist AutomaticallyInstallMacOSUpdates -bool false
+    sudo defaults write /Library/Preferences/com.apple.SoftwareUpdate.plist CriticalUpdateInstall -bool false
+    sudo defaults write /Library/Preferences/com.apple.SoftwareUpdate.plist ConfigDataInstall -bool false
```
This does not seem to work for me. These are the settings that I see after running the script (I had intentionally enabled updates in the macOS setup wizard):

<img width="1288" height="893" alt="Screenshot 2025-10-03 at 12 35 37 PM" src="https://github.com/user-attachments/assets/cd54ddbb-d916-4d31-bece-2127cce5360e" />

Does this seem correct?

celanthe replied:
No, I see the issue and will correct in a new commit.

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,217 @@
+cleanup_system() {
+    log "Cleaning up system..."
+    
+    # Close all applications (except essential system apps)
+    osascript -e 'tell application "System Events" to set quitapps to name of every application process whose background only is false'
+    osascript -e 'repeat with apps in quitapps
+        if apps is not in {"Finder", "loginwindow"} then
+            tell application apps to quit
+        end if
+    end repeat' 2>/dev/null || true
```
This does not work for me:

<img width="1288" height="893" alt="Screenshot 2025-10-03 at 12 25 38 PM" src="https://github.com/user-attachments/assets/b6942e5c-b1df-4a44-819e-4bd02c490ecc" />

After the script completes and system reboots, Terminal, Safari and System Preferences are all still opened

celanthe replied:
TY for the catch, will push a new commit to hopefully fix. Terminal won't close until it's done running the process, but the others should.

spikeburton commented on README.md:
````diff
@@ -50,27 +54,21 @@ To deploy a Ventura VM with SIP disabled, run:
 1. Select 'Pull from OCI registry'
 1. Name VM and set parameters (CPUs, Memory, HD size)
 1. In the 'OCI Image Name' field enter: `ghcr.io/macstadium/orka-images/tahoe:latest`
-```sh  
-  ghcr.io/macstadium/orka-images/tahoe:latest
-```
````
Why remove this btw? It allows you to copy it with a simple button from the GitHub UI which is handy

Same with the one below

celanthe replied:
Will correct.

celanthe committed:
Update setup/README.md

Co-authored-by: Spike Burton <44298491+spikeburton@users.noreply.github.com>

celanthe committed:
fix: update script and README

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,202 @@
+
+# Enable SSH
+
+sudo launchctl load -w /System/Library/LaunchDaemons/ssh.plist 2>/dev/null || true
```
This will run before `main` does now (we've removed the function this was wrapped in)

celanthe replied:
fixed in a new commit

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,202 @@
+    
+    # Enable required services
+    enable_screen_sharing
+    enable_remote_login
```
This will fail since `enable_remote_login` no longer exists

celanthe replied:
fixed in a new commit

celanthe committed:
fix: enable remote login logic

celanthe committed:
fix: remove superflouous comments

celanthe committed:
fix: remove more comments

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,183 @@
+enable_remote_login() {
+    log "Enabling Remote Login (SSH)..."
+   
+    sudo systemsetup -setremotelogin on
+    
+    sudo launchctl load -w /System/Library/LaunchDaemons/ssh.plist 2>/dev/null || true
```
I see that `sudo systemsetup -setremotelogin on` is back. Didn't we decide to remove this?

Note - we should only need 1 or the other - either the `systemsetup` command or the `launchctl` command. Not both - it is redundant

I would just use `launchctl`. As `systemsetup` here requires Full Disk Access.

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,183 @@
+        end try
+    end if
+end repeat
+EOF
```
This does not work. It asks to kill Terminal - which also kills the running script (see the popup)

<img width="1920" height="1080" alt="Screenshot 2025-10-03 at 3 59 52 PM" src="https://github.com/user-attachments/assets/8f63799e-5fc0-4306-82ec-8b121b8a90bd" />

Do we really need any of the logic in this `cleanup_system` function? Note that removing from Downloads will also cause an alert to popup as you must grant Terminal access to this directory explicitly

I would opt to just keep it simple. And instruct users to just close open apps before pushing their image

celanthe replied:
The cleanup_system function mimics the instructions we have in the image setup instructions which are to delete the script from the downloads folder, clear the terminal history, etc. I can remove all of that and just make a note in the docs, sure.

spikeburton replied:
Is there a script in the Downloads folder now? Just clearing `/tmp` seems sufficient :)) Clearing the terminal history we can keep as well, but not mandatory.

In general - our script should not be too opinionated. The real goal is to
1. Make sure SSH / Screen Share are enabled
2. Install our VM tools
3. Run the launch daemon to improve screen sharing performance (not strictly required, but good to have)

Everything else is really up to the user how they wish to configure the guest OS - the above are what are required for the VM to work with Orka

celanthe replied:
Removed the terminal deletion section in the most recent commit

celanthe committed:
fix: remove unnecessary cleanup steps, update README

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,163 @@
+}
+
+# Configure macOS updates
+configure_macos_updates() {
```
This still does not work for me btw. Updates are still enabled if I check in System Preferences after running the script.

If we cannot figure this out - I would suggest just to remove it. As it is not strictly required

celanthe replied:
Removed this step, will review the readme and update the docs to clarify system updates etc. for Tahoe

celanthe committed:
remove terminal history and system updates

removed terminal history deletion step and system updates step

celanthe committed:
Update README.md

adds clarification for users running MacOS Tahoe

spikeburton commented on setup/README.md:
The script is looking pretty good. I will test it one more time tomorrow.

Can we go through the README and make sure that all information is accurate and reflects the current state of the script? And make any updates as needed

celanthe replied:
Sure, I'll go through this today and make any changes that need to be made.

spikeburton commented:
It seems like something is not working properly with Screen Sharing.

I created a new VM with Desktop, ran the script, and pushed it to the Harbor instance in my cluster.

I then deployed a VM using the image with `orka3`. I can connect to the VM just fine with SSH (this is working), but Screen Share does not seem to work.

<img width="969" height="452" alt="Screenshot 2025-10-07 at 12 51 54 PM" src="https://github.com/user-attachments/assets/1ba49e4d-7cad-4772-96c6-28999e02e7dc" />

If I SSH to the VM, I see that the screen sharing service is not running:

```
spikeburton@Spikes-Virtual-Machine ~ % sudo launchctl print system/com.apple.screensharing
system/com.apple.screensharing = {
        active count = 0
        path = /System/Library/LaunchDaemons/com.apple.screensharing.plist
        type = LaunchDaemon
        state = not running

        program = /System/Library/CoreServices/RemoteManagement/screensharingd.bundle/Contents/MacOS/screensharingd
        arguments = {
                /System/Library/CoreServices/RemoteManagement/screensharingd.bundle/Contents/MacOS/screensharingd
        }

        default environment = {
                PATH => /usr/bin:/bin:/usr/sbin:/sbin
        }

        environment = {
                OSLogRateLimit => 64
                MallocSpaceEfficient => 1
                XPC_SERVICE_NAME => com.apple.screensharing
        }

        domain = system
        minimum runtime = 10
        base minimum runtime = 10
        exit timeout = 5
        runs = 1
        last exit code = 0

        event triggers = {
                Listener => {
                        keepalive = 0                                                                                             service = com.apple.screensharing
                        stream = com.apple.bonjour.registration
                        monitor = com.apple.UserEventAgent-System
                        descriptor = {
                                "Bonjour" => "rfb"
                                "SockServiceName" => "vnc-server"
                        }
                }
        }

        endpoints = {
                "com.apple.screensharing.server" = {
                        port = 0x0
                        active = 0
                        managed = 1
                        reset = 1
                        hide = 1
                        watching = 1
                }
        }

        event channels = {
                "com.apple.bonjour.registration" = {
                        port = 0xa0f
                        active = 0
                        managed = 1
                        reset = 0
                        hide = 0
                        watching = 1
                }
        }

        sockets = {
                "Listener" = {
                        type = stream
                        service name = vnc-server

                        sockets = {
                                15 (no bytes to read)
                                16 (no bytes to read)
                        }

                        active = 0
                        passive = 1
                        bonjour = 1
                        ipv4v6 = 0
                        receive_packet_info = 0
                }
        }

        resource coalition = {
                ID = 381
                type = resource
                state = active
                active count = 1
                name = com.apple.screensharing
        }

        jetsam coalition = {
                ID = 382
                type = jetsam
                state = active
                active count = 1
                name = com.apple.screensharing
        }

        spawn type = interactive (4)
        jetsam priority = 40
        jetsam memory limit (active, soft) = 1000 MB
        jetsam memory limit (inactive, soft) = 1000 MB
        jetsamproperties category = daemon
        jetsam thread limit = 32
        cpumon = default
        exponential throttling grace limit = 10
        job state = exited
                                                                                                                  properties = partial import | supports transactions | inferred program | system service | exponential throttling | tle system
}
```

This we should look into

celanthe commented:
> It seems like something is not working properly with Screen Sharing.
> 
> I created a new VM with Desktop, ran the script, and pushed it to the Harbor instance in my cluster.
> 
> I then deployed a VM using the image with `orka3`. I can connect to the VM just fine with SSH (this is working), but Screen Share does not seem to work.
> 
> <img alt="Screenshot 2025-10-07 at 12 51 54 PM" width="969" height="452" src="https://private-user-images.githubusercontent.com/44298491/498420202-1ba49e4d-7cad-4772-96c6-28999e02e7dc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk4NjE3NzcsIm5iZiI6MTc1OTg2MTQ3NywicGF0aCI6Ii80NDI5ODQ5MS80OTg0MjAyMDItMWJhNDllNGQtN2NhZC00NzcyLTk2YzYtMjg5OTllMDJlN2RjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA3VDE4MjQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2MjIxNTA3YjU3ZDg5OWQyNDJjYmFkZTM3MTc4ZmI1YzQ2NTFjMmNkMjAzMDA1MzlkNTQxMGQwNTdhMjUwNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8xCmCAXWv1L1rejdu6aRyAqD-t6Ezw9jRGaFCxDI11o">
> If I SSH to the VM, I see that the screen sharing service is not running:
> 
> ```
> spikeburton@Spikes-Virtual-Machine ~ % sudo launchctl print system/com.apple.screensharing
> system/com.apple.screensharing = {
>         active count = 0
>         path = /System/Library/LaunchDaemons/com.apple.screensharing.plist
>         type = LaunchDaemon
>         state = not running
> 
>         program = /System/Library/CoreServices/RemoteManagement/screensharingd.bundle/Contents/MacOS/screensharingd
>         arguments = {
>                 /System/Library/CoreServices/RemoteManagement/screensharingd.bundle/Contents/MacOS/screensharingd
>         }
> 
>         default environment = {
>                 PATH => /usr/bin:/bin:/usr/sbin:/sbin
>         }
> 
>         environment = {
>                 OSLogRateLimit => 64
>                 MallocSpaceEfficient => 1
>                 XPC_SERVICE_NAME => com.apple.screensharing
>         }
> 
>         domain = system
>         minimum runtime = 10
>         base minimum runtime = 10
>         exit timeout = 5
>         runs = 1
>         last exit code = 0
> 
>         event triggers = {
>                 Listener => {
>                         keepalive = 0                                                                                             service = com.apple.screensharing
>                         stream = com.apple.bonjour.registration
>                         monitor = com.apple.UserEventAgent-System
>                         descriptor = {
>                                 "Bonjour" => "rfb"
>                                 "SockServiceName" => "vnc-server"
>                         }
>                 }
>         }
> 
>         endpoints = {
>                 "com.apple.screensharing.server" = {
>                         port = 0x0
>                         active = 0
>                         managed = 1
>                         reset = 1
>                         hide = 1
>                         watching = 1
>                 }
>         }
> 
>         event channels = {
>                 "com.apple.bonjour.registration" = {
>                         port = 0xa0f
>                         active = 0
>                         managed = 1
>                         reset = 0
>                         hide = 0
>                         watching = 1
>                 }
>         }
> 
>         sockets = {
>                 "Listener" = {
>                         type = stream
>                         service name = vnc-server
> 
>                         sockets = {
>                                 15 (no bytes to read)
>                                 16 (no bytes to read)
>                         }
> 
>                         active = 0
>                         passive = 1
>                         bonjour = 1
>                         ipv4v6 = 0
>                         receive_packet_info = 0
>                 }
>         }
> 
>         resource coalition = {
>                 ID = 381
>                 type = resource
>                 state = active
>                 active count = 1
>                 name = com.apple.screensharing
>         }
> 
>         jetsam coalition = {
>                 ID = 382
>                 type = jetsam
>                 state = active
>                 active count = 1
>                 name = com.apple.screensharing
>         }
> 
>         spawn type = interactive (4)
>         jetsam priority = 40
>         jetsam memory limit (active, soft) = 1000 MB
>         jetsam memory limit (inactive, soft) = 1000 MB
>         jetsamproperties category = daemon
>         jetsam thread limit = 32
>         cpumon = default
>         exponential throttling grace limit = 10
>         job state = exited
>                                                                                                                   properties = partial import | supports transactions | inferred program | system service | exponential throttling | tle system
> }
> ```
> 
> This we should look into

Will look into this thanks

celanthe committed:
fix: add kickstart step to fully enable screen sharing

celanthe committed:
fix: clean up script

celanthe committed:
fix: update script

celanthe committed:
fix: update screen sharing step

celanthe committed:
fix: update screen sharing step

celanthe committed:
fix: update setup script

celanthe committed:
fix: update screen sharing step

celanthe committed:
fix: remove conflicting commands, add error messages

celanthe committed:
Update README.md

Add clarification that one will need to toggle screen sharing off and then back on for settings to persist when connecting via SSH

celanthe committed:
fix: update script and README for clarity and remove unnecessary automation

spikeburton commented on setup/setup.sh:
```diff
@@ -31,32 +38,22 @@ if ! command -v sudo &> /dev/null; then
 fi
 
-log "Starting MacOS Orka VM setup..."
+ensure_autologin() {
```
Why do we need to ensure autologin?

celanthe replied:
removed this

spikeburton commented on setup/setup.sh:
```diff
@@ -84,44 +80,143 @@ setup_sys_daemon() {
     
-    sudo shutdown -r +0.5 "Rebooting to flush VM changes" 2>/dev/null || \
-    sudo reboot
+    local filevault_status
```
Do we need the FileVault check?

celanthe replied:
no, removed it

spikeburton commented on setup/setup.sh:
```diff
@@ -84,44 +80,143 @@ setup_sys_daemon() {
+    echo "3. GRANT FULL DISK ACCESS TO SSH:"
+    echo "   - Open System Settings > Privacy & Security > Full Disk Access"
+    echo "   - Click the '+' button"
+    echo "   - Press Cmd+Shift+G and enter: /usr/libexec/sshd-keygen-wrapper"
```
This is not required to be done manually btw.

All that is required is:
* Enable System Preferences -> General -> Sharing - > Remote Login (this will enable SSH)
* Click the ⓘ (info button) and select "Allow Full Disk Access for remote users"

This will add full disk access for the `sshd-keygen-wrapper` helper

celanthe replied:
removed this and updated the docs to refelect

spikeburton reviewed (commented):
We have now changed a few things. A few general comments:
* Why do we need any logic around auto login at all?
* Do we need the manual steps printed from the script?

I would have the manual steps mentioned in the README with screen shots. This will be way more useful.

I would again suggest keeping the script as simple as possible:
* Install our VM tools
* Run the `setup-sys-daemon.sh` script

These are the important things that we can easily automate. I am not completely against printing some helpful information when running the script, but I would keep it as concise as possible and direct users to the README for the manual steps.

celanthe committed:
fix: update script and docs, add screenshots

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+### Before running the script
+
+- Ensure you have administrator access to the VM
+- Close any important applications (they will be terminated during cleanup)
```
This is no longer true, right?

celanthe replied:
Correct, updated the README.

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+
+- Ensure you have administrator access to the VM
+- Close any important applications (they will be terminated during cleanup)
+- Toggle 'Screen Sharing' off, then back on via the GUI **System Preferences -> General -> Sharing -> Screen Sharing** so that system settings persist when connecting to a VM via SSH.
```
Why toggle "off"? This should not be needed. It just needs to be enabled, right?

celanthe replied:
Resolved in a new commit

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+- Close any important applications (they will be terminated during cleanup)
+- Toggle 'Screen Sharing' off, then back on via the GUI **System Preferences -> General -> Sharing -> Screen Sharing** so that system settings persist when connecting to a VM via SSH.
+![Screen sharing toggled 'on' in system settings](./assets/screensharing.png)
+- Toggle 'Remote Login' off, then back on via the GUI  **System Preferences -> General -> Sharing -> Remote Login**
```
Same for SSH

celanthe replied:
Updated this

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+![Screen sharing toggled 'on' in system settings](./assets/screensharing.png)
+- Toggle 'Remote Login' off, then back on via the GUI  **System Preferences -> General -> Sharing -> Remote Login**
+![Remote login enabled in system settings](./assets/remotelogin.png)
+- Enable 'Full Disk Access' to SSH via **System Settings > Privacy & Security > Full Disk Access** -> Click the "+" button
```
This is not quite right.

You just need to go to System Preferences -> General -> Sharing and click the ⓘ icon next to Remote Login, then select Full disk access for remote users (like in the screen shot we have now)

celanthe replied:
Updated!

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+
+#### On MacOS Sequoia
+
+- When prompted for Terminal permissions access, click 'Accept'
```
Is this true now?

celanthe replied:
Yes, it is still true for me when I run the script. Verified it across OSes.

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+- When prompted for Terminal permissions access, click 'Accept'
+- Enter your password when prompted during the script installation process
+- When the script shell displays `sh-3.2#` type `exit` and press `Enter` to continue script installation
+- Upon system reboot, re-open the Terminal application and enter `sudo launchctl list sysctl` to confirm the Orka sys-daemon script has installed. You should see an output similar to:
```
This is a nice sanity check, but not strictly required

Note - this is somewhat of a nitpick

celanthe replied:
updated the README to note this step is optional

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,82 @@
+
+### Important notes
+
+- Change default passwords in production environments
```
Are there default passwords we provide?

celanthe replied:
removed this in a new commit

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,97 @@
+    log "Performing minimal system cleanup..."
+    
+    rm -rf /tmp/* 2>/dev/null || true
+    history -c 2>/dev/null || true
```
btw - this might be an unintended (and undesirable) side effect - to clear the terminal history.

As in - it might be unexpected for someone who runs this script

celanthe replied:
removed this in a new commit

spikeburton reviewed (commented):
I have several questions about the README. It still looks like it has not been updated after we have changed the script. Can we look through it for accuracy? As the documentation is important

celanthe committed:
fix: remove cleanup step from script, update README

celanthe committed:
fix: update README for clarity

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,64 @@
+### Before running the script
+
+- Toggle 'Screen Sharing' on via the GUI **System Preferences -> General -> Sharing -> Screen Sharing** so that system settings persist when connecting to a VM via SSH.
+![Screen sharing toggled 'on' in system settings](./assets/screensharing.png)
```
None of the images render for me btw. It appears that the path is incorrect. The `assets` folder is in the parent directory, not the current directory

btw - have you previewed the README? If you use VS Code, the built in markdown renderer is pretty good 🙂 You can even preview while you are editing the file with `CMD+K V`

I would highly suggest previewing the work before submitting it for review and make sure that it looks good to you first.

celanthe replied:
I do use VS Code. Thanks, I'll fix this up, and yes I do know how to preview markdown, appreciate the reminder! Will fix the image path.

spikeburton commented on setup/README.md:
```diff
@@ -0,0 +1,64 @@
+![Screen sharing toggled 'on' in system settings](./assets/screensharing.png)
+- Toggle 'Remote Login' on via the GUI  **System Preferences -> General -> Sharing -> Remote Login**
+![Remote login enabled in system settings](./assets/remotelogin.png)
+- Enable 'Full Disk Access' to SSH via the GUI **System Settings -> General -> Sharing**, click the ⓘ icon next to "Remote Login", and select 'Allow full disk access for remote users'
```
Something I just noticed btw - if you enable SSH via System Preferences, Full Disk Access for remote users is enabled by default. It doesn't hurt to mention it here of course :))

celanthe replied:
Awesome, I'll make a note or update this to clarify

spikeburton reviewed (commented):
The script is working nicely. There is one issue with the README that needs to be addressed (for the screen shots). Other than this - it is looking pretty good

celanthe committed:
fix: update image path + clarify step for full disk access

spikeburton commented on setup/setup.sh:
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+
+ORKA_VM_TOOLS_VERSION="${ORKA_VM_TOOLS_VERSION:-3.5.0}"
+CURRENT_USER="${USER}"
```
`nit` - this variable is not used, is it?
